### PR TITLE
[python] Append-mode I/O step

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:**

#1278

**Changes:**

Leverage #1554 and #1556, then connect up to data ingestion per se.

Basic use-case which is unit-tested in this PR, and included in on-line help on this PR, and will additionally be illustrated by example-notebook material as well as an upcoming [TileDB-Cloud-Py](https://github.com/TileDB-Inc/TileDB-Cloud-Py) PR:

```
# This step must be sequential
rd = tiledbsoma.io.register(
    experiment_uri,
    h5ad_file_names,
    measurement_name=measurement_name,
    obs_field_name="obs_id",
    var_field_name="var_id",
    context=context,
)

# This loop can be parallelized over h5ad_file_names
for h5ad_file_name in h5ad_file_names:
    tiledbsoma.io.from_h5ad(
        experiment_uri,
        h5ad_file_name,
        measurement_name=measurement_name,
        registration_mapping=rd,
    )
```

**Notes for Reviewer:**

* Another PR will be necessary after an audit of how good/useful the logging is.

### CODE REVIEW

**Comment by johnkerl** [comment]
Ready for review _with caveats_:

* Some on-line help bits marked `XXX` need to be written
* A couple fast-fail checks need to get hoisted to top of method

Feel free to ignore this PR until I implement those bits (I'll ping here when they're completed) -- or, feel free to peek sooner @aaronwolen as the essentials of this PR --- especially as walked through by unit-test cases --- is ready to go.

**Comment by johnkerl** [comment]
@thetorpedodog @aaronwolen re
https://github.com/single-cell-data/TileDB-SOMA/pull/1587#issuecomment-1677928905
those caveats have been addressed.

However do please see the description field of this PR for caveats on which things this particular PR does and does not address.

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 1]
```diff
@@ -342,6 +390,9 @@ def from_anndata(
                 platform_config=platform_config,
                 context=context,
                 ingestion_params=ingestion_params,
+                # XXX CHECK EXISTENCE OF THE LAYER -- FATAL IF NOT PRESENT
+                # XXX HAVE CHECKED THIS EARLY FOR FAIL-FAST
```
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/io/ingest.py:394]: Is this something you wanted to address now or is this a TODO for later?

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 2]
```diff
@@ -649,6 +717,50 @@ def _create_or_open_coll(
...
+    _new_ genes not yet seen before (if any).
+    """
+    try:
+        with _factory.open(
+            df_uri, "r", soma_type=DataFrame, context=context
+        ) as previous_soma_dataframe:
+            previous_table = previous_soma_dataframe.read().concat()
+            previous_df = previous_table.to_pandas()
+            previous_join_ids = set(
+                list(int(e) for e in get_dataframe_values(previous_df, SOMA_JOINID))
```
Comment (thetorpedodog) [inline apis/python/src/tiledbsoma/io/ingest.py:754]: this can just be `set(int(e) for e in ...)`

**Review by thetorpedodog — COMMENTED** [review]
I see no major issues, just a couple small things to point out.

File: apis/python/src/tiledbsoma/io/ingest.py (hunk 1 above)
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/ingest.py:394]: @thetorpedodog I am addressing that now & had worked off the to-do and removed the XXX comment.

Let me double-check my work.

Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/ingest.py]: OK sorry

Just re-pushed

**Comment by johnkerl** [comment]
@aaronwolen @thetorpedodog ping 🙏

**Comment by johnkerl** [comment]
@aaronwolen @thetorpedodog ping 🙏

**Comment by johnkerl** [comment]
@aaronwolen @thetorpedodog ping 🙏

**Review by aaronwolen — APPROVED** [review]
🚀

**Pull request merged**